src/input/key-state.c

```diff
@@ -69,20 +69,32 @@ key_state_nr_pressed_sent_keycodes(void)
 }
 
 void
-key_state_set_pressed(uint32_t keycode, bool ispressed)
+key_state_set_pressed(uint32_t keycode, bool is_pressed, bool is_modifier)
 {
-	if (ispressed) {
+	if (is_pressed) {
 		add_key(&pressed, keycode);
+		if (is_modifier) {
+			add_key(&pressed_mods, keycode);
+		}
 	} else {
 		remove_key(&pressed, keycode);
+		remove_key(&pressed_mods, keycode);
 	}
 }
 
 void
-key_state_store_pressed_keys_as_bound(void)
+key_state_store_pressed_key_as_bound(uint32_t keycode)
 {
-	memcpy(bound.keys, pressed.keys, MAX_PRESSED_KEYS * sizeof(uint32_t));
-	bound.nr_keys = pressed.nr_keys;
+	add_key(&bound, keycode);
+	/*
+	 * Also store any pressed modifiers as bound. This prevents
+	 * applications from seeing and handling the release event for
+	 * a modifier key that was part of a keybinding (e.g. Firefox
+	 * displays its menu bar for a lone Alt press + release).
+	 */
+	for (int i = 0; i < pressed_mods.nr_keys; ++i) {
+		add_key(&bound, pressed_mods.keys[i]);
+	}
```

Comment:
Just wanted to point out something: In a `W-Left` example,
- We'd send a press event for the `Super` key to the focused surfaces (`Super_L` or `Super_R`).
- Once the `m` comes in we'd then handle they keybind
- Store the currently set of pressed keys and modifiers as bound
- And thus don't send the release for the earlier `Super_L` / `Super_R`, resulting in the client getting a stray `Press` for the modifier key itself

We can't really prevent the `Super_L` / `Super_R` key to be sent to the application because at that point we don't yet know if it is actually a part of a keybind or not. But I guess we should try to make sure not the eat its release *if* it has already been sent. Maybe applications are also fine with `Press` events for modifiers, I don't know. It feels wrong though unless I miss something.

As we currently do not store the modifiers as pressed this would not happen and the application would see the corresponding `Release` event.

Another thing: the `is_modifier` state may be wrong in case of layout switching, e.g. it might show up as that group_next thing and thus end up in the usual pressed array, depending on the order of Press and Release events for the layout switch.

Reply:
> We can't really prevent the `Super_L` / `Super_R` key to be sent to the application because at that point we don't yet know if it is actually a part of a keybind or not. But I guess we should try to make sure not the eat its release _if_ it has already been sent. Maybe applications are also fine with `Press` events for modifiers, I don't know. It feels wrong though unless I miss something.

I've never seen this documented anywhere, but believe applications cope fine with seeing modifier press without associated release. And we've done it like that for a long time. I think `alt` is the example the convinced me... (press+release triggers menu). IIRC the thing the makes applications misbehave is sending a release without having first sent a press.
 
> As we currently do not store the modifiers as pressed this would not happen and the application would see the corresponding `Release` event.

Well - until recently they used all (modifiers + non-modifiers) be stored in the pressed array :smile:

> Another thing: the `is_modifier` state may be wrong in case of layout switching, e.g. it might show up as that group_next thing and thus end up in the usual pressed array, depending on the order of Press and Release events for the layout switch.

I tried to reproduce that, but seems to have gone away (i.e. press shift + press caps + release shift + release caps).

I can't test the keybind with `XF86LaunchA` thing (issue #1223). @jp7677 any chance of testing this branch against that?

Reply:
> We can't really prevent the Super_L / Super_R key to be sent to the application because at that point we don't yet know if it is actually a part of a keybind or not. But I guess we should try to make sure not to eat its release if it has already been sent. Maybe applications are also fine with stray Press events for modifiers, I don't know. It feels wrong though unless I miss something.

This initially felt wrong to me too. But it's existing behavior (in both 0.6.5 and master) at least in the Alt-Tab case -- we eat the Alt release-event. And it seems to be the only way to prevent e.g. Firefox from seeing a lone Alt press + release and displaying its menu bar.

> Another thing: the is_modifier state may be wrong in case of layout switching, e.g. it might show up as that group_next thing and thus end up in the usual pressed array, depending on the order of Press and Release events for the layout switch. The corresponding Release may then show up correctly as a modifier. Or the other way around.

This should be handled reasonably well. Both modifiers and non-modifiers are stored in the "pressed" array unconditionally (as raw keycodes). And on release, keycodes are removed from both "pressed" and "pressed_mods" unconditionally. So the only consequence would be in the case where the press event is not detected as a modifier -- then it will not be stored in the "bound" array. I don't think there is much we can do about that.